pkg/discovery/transport/client/client.go

```diff
@@ -140,14 +157,20 @@ func (c *Client) Stop() error {
 func (c *Client) streamOut() error {
 	for {
 		select {
+		case <-c.conf.Context.Done():
+			c.conf.Logger.Debug("Closing streamOut as context is done.")
+			return nil
 		case ev := <-c.conf.Out:
+			c.conf.Logger.Debugf("Sending event %v", ev)
 			err := c.stream.Send(ev)
 			if err != nil {
-				c.conf.ErrCh <- err
+				c.conf.Logger.Errorf("Closing streamOut as error occurred: %v", err)
+				select {
+				case c.conf.ErrCh <- err:
```

Comment:
Added comment:
```go
// The ErrCh is a buffered channel shared by multiple subroutines. Any error written to the channel
// indicates that the current procedure has failed.
// While the "root" error is sufficient to indicate that the routine failed, it may cause further 
// errors in other routines. If write to ErrCh fails, err is classified as a consequent error. In 
// this case, "err" is discarded to prevent the routine from blocking.
```